### DISCUSSION TITLE
Suggestions for providing Copilot more powerful context / prompts

### DISCUSSION THREAD

**Discussion by cashgarman** [community/community#7142] [Copilot Conversations]
I've been using Copilot quite heavily in both production and non-production code for a couple of weeks now and I have to, first of all, say it's changed my outlook on where my career will be over the coming years if not months. Fantastic work to the no doubt hundreds of talented people involved in this bright new future :)

That being said, I think like anyone else who has been using Copilot in a serious way for a while will have also noticed, you develop techniques and tactics to help give it the best chance to predict what would be the best suggestion in the current context.

One of my techniques has been to use what I'm calling a 'context puck', which is effectively a large block comment I populate with relevant code from other files directly above the area I'm working on. As I move down a file or class, I bring the puck with me and update it with relevant code as I go. I've found this incredibly effective at improving suggestions.

That being said, it's fairly labour intensive and can interrupt your flow a fair bit.

So I've been thinking of possible automated ways of helping streamline this. Most of which I imagine could be implemented at the IDE plugin level, and not in Copilot itself.

So I guess here is a wish list of features I would love to see:

1. As opposed to maintaining a floating lump of commented out context, it would be great if you could split your IDE editor view vertically or horizontally, and any highlighted code in the opposite pane(s) to the one you are working in is injected into the context given to Copilot for its generation. This would allow for very targeting prompting using code or even documentation from anywhere in your project.

2. It would be great if Copilot respected folding/collapsed code. As in it ignored it when gathering its context for the generation. I've done some testing and this definitely doesn't seem to be the case at the moment but I feel it would be incredibly useful as you could leave very detailed but folded/collapsed comments above every class/method without interfering with Copilot's maximum vertical limit (which seems to cap out around a couple of hundred lines of code if I'm not mistaken).

3. It would be useful if, at least as an option, you could enable downwards travel on the gathering of context. This definitely does not seem to be the case at the moment. But it would allow you to avoid the 'chicken and egg' problem of having to sometimes juggle 2 classes/methods vertically that are very interdependent on each other to maintain the most relevant context. This would also be incredibly useful in letting Copilot suggest a comment for the immediately following lines of code, as it might help uncover if your code is actually being 'perceived' to do what you think it is, great for sanity checking. This one I admit might be pushing at the limits of much the ML is trained/works.

4. Being able to accept a suggestion 1 token at a time as opposed to the entire suggestion would be fantastic. With a keyboard shortcut to slowly work your way through a large suggestion, making corrections as you go.

5. I suspect this might already be in the works, but being able to manually trigger a re-suggestion of either the current line of code or a selected block of code would be great.

If anyone has any other suggestions or tips and tricks they've found to help them work most effectively with Copilot, I'd love to hear them :)

Thanks again Copilot team for your inspiring work, I really look forward to the future of AI programming :)

**Comment by douglasg14b** [comment]
Really what we need is https://githubnext.com/projects/copilot-view/

A personal project of mine has nearly 1000 files, and often changes are not limited to just one file or area, and involve adding boilerplate in multiple areas. Or hunting down a utility, feature, or other such artifact that solves a problem I'm dealing with but I can't remember if this has already been solved in the codebase. If copilot knew about the entire codebase, it could make better suggestions, and even suggest multi-file changes that follow the patterns & idioms of your codebase.

Or maybe a premium-like feature that lets us have models trained for specific code bases
